File: cs/Minsk/CodeAnalysis/Syntax/Lexer.cs
using System.Collections;
using Minsk.CodeAnalysis.Text;

namespace Minsk.CodeAnalysis.Syntax;

internal sealed class Lexer : IEnumerable<SyntaxToken>
{
    private readonly DiagnosticBag _diagnostics = new();
    private readonly SourceText _text;
    private int _position;

    public Lexer(SourceText text)
    {
        _text = text;
    }

    public IEnumerable<Diagnostic> Diagnostics => _diagnostics;

    private char CurrentChar => Peek(0);

    public IEnumerator<SyntaxToken> GetEnumerator()
    {
        while (true)
        {
            var start = _position;
            var kind = SyntaxKind.BadToken;
            string? currentText = null;
            object? value = null;

            if (char.IsWhiteSpace(CurrentChar))
            {
                while (char.IsWhiteSpace(CurrentChar))
                {
                    _position++;
                }

                kind = SyntaxKind.WhitespaceToken;
            }
            else if (char.IsDigit(CurrentChar))
            {
                while (char.IsDigit(CurrentChar))
                {
                    _position++;
                }

                currentText = CurrentText(start);
                if (!int.TryParse(currentText, out var intVal))
                {
                    _diagnostics.ReportInvalidInt(new TextSpan(start, _position - start), currentText, typeof(int));
                }

                kind = SyntaxKind.NumberToken;
                value = intVal;
            }
            else if (char.IsLetter(CurrentChar))
            {
                while (char.IsLetterOrDigit(CurrentChar))
                {
                    _position++;
                }

                currentText = CurrentText(start);
                kind = SyntaxFacts.GetKeywordKind(currentText);
            }
            else
            {
                switch (CurrentChar)
                {
                    case '+':
                        kind = SyntaxKind.PlusToken;
                        _position++;
                        break;
                    case '-':
                        kind = SyntaxKind.MinusToken;
                        _position++;
                        break;
                    case '*':
                        kind = SyntaxKind.StarToken;
                        _position++;
                        break;
                    case '/':
                        kind = SyntaxKind.SlashToken;
                        _position++;
                        break;
                    case '(':
                        kind = SyntaxKind.OpenParenthesisToken;
                        _position++;
                        break;
                    case '{':
                        kind = SyntaxKind.OpenBraceToken;
                        _position++;
                        break;
                    case '}':
                        kind = SyntaxKind.CloseBraceToken;
                        _position++;
                        break;
                    case ')':
                        kind = SyntaxKind.CloseParenthesisToken;
                        _position++;
                        break;
                    case '!':
                        if (Peek(1) == '=')
                        {
                            kind = SyntaxKind.BangEqualsToken;
                            _position += 2;
                        }
                        else
                        {
                            kind = SyntaxKind.BangToken;
                            _position++;
                        }

                        break;
                    case '=':
                        if (Peek(1) == '=')
                        {
                            kind = SyntaxKind.EqualsEqualsToken;
                            _position += 2;
                        }
                        else
                        {
                            kind = SyntaxKind.EqualsToken;
                            _position++;
                        }

                        break;
                    case '<':
                        if (Peek(1) == '=')
                        {
                            kind = SyntaxKind.LessEqualsToken;
                            _position += 2;
                        }
                        else
                        {
                            kind = SyntaxKind.LessToken;
                            _position++;
                        }

                        break;
                    case '>':
                        if (Peek(1) == '=')
                        {
                            kind = SyntaxKind.GreaterEqualsToken;
                            _position += 2;
                        }
                        else
                        {
                            kind = SyntaxKind.GreaterToken;
                            _position++;
                        }

                        break;
                    case '&':
                        if (Peek(1) == '&')
                        {
                            kind = SyntaxKind.AmpersandAmpersandToken;
                            _position += 2;
                        }

                        break;
                    case '|':
                        if (Peek(1) == '|')
                        {
                            kind = SyntaxKind.PipePipeToken;
                            _position += 2;
                        }

                        break;
                    case '\0':
                        yield break;
                }
            }

            if (kind == SyntaxKind.BadToken)
            {
                _diagnostics.ReportBadCharacter(_position, CurrentChar);
                _position++;
            }

            yield return new SyntaxToken(kind, currentText ?? CurrentText(start), start, value);
        }
    }

    IEnumerator IEnumerable.GetEnumerator()
    {
        return GetEnumerator();
    }

    private char Peek(int offset)
    {
        var index = _position + offset;
        if (index >= _text.Length)
        {
            return '\0';
        }

        return _text[index];
    }

    private string CurrentText(int start)
    {
        return _text.ToString(start, _position - start);
    }
}

File: nim/src/minsk/codeAnalysis/syntax/lexer.nim
import std/strformat
import std/strutils

import minsk/minskObject

import syntaxFacts
import syntaxKind
import syntaxToken

type
  Lexer* = object
    text: string
    position: int
    mDiagnostics: seq[string]

proc newLexer*(text: string): Lexer =
  result.text = text
  result.position = 0
  result.mDiagnostics = @[]

proc peek(lexer: Lexer, offset: int): char =
  let index = lexer.position + offset
  if index >= lexer.text.len:
    return '\0'
  lexer.text[index]

proc current(lexer: Lexer): char =
  lexer.peek(0)

proc nextToken*(lexer: var Lexer): SyntaxToken =
  if lexer.position >= lexer.text.len:
    return newToken(
      SyntaxKind.EndOfFileToken,
      lexer.position,
      "",
      moNull(),
    )
  let start = lexer.position
  var kind = SyntaxKind.BadToken
  var text = ""
  var value = moNull()
  if lexer.current.isDigit:
    while lexer.current.isDigit:
      lexer.position += 1
    text = lexer.text[start..<lexer.position]
    try:
      value = moInteger(text.parseInt())
    except ValueError:
      lexer.mDiagnostics.add(fmt"The number {text} isn't a valid int.")
    kind = SyntaxKind.NumberToken
  elif lexer.current.isSpaceAscii:
    while lexer.current.isSpaceAscii:
      lexer.position += 1
    text = lexer.text[start..<lexer.position]
    kind = SyntaxKind.WhitespaceToken
  elif lexer.current.isAlphaAscii:
    while lexer.current.isAlphaAscii:
      lexer.position += 1
    text = lexer.text[start..<lexer.position]
    kind = keywordKind(text)
  else:
    case lexer.current
    of '+':
      lexer.position += 1
      kind = SyntaxKind.PlusToken
    of '-':
      lexer.position += 1
      kind = SyntaxKind.MinusToken
    of '*':
      lexer.position += 1
      kind = SyntaxKind.StarToken
    of '/':
      lexer.position += 1
      kind = SyntaxKind.SlashToken
    of '(':
      lexer.position += 1
      kind = SyntaxKind.OpenParenthesisToken
    of ')':
      lexer.position += 1
      kind = SyntaxKind.CloseParenthesisToken
    of '!':
      if lexer.peek(1) == '=':
        lexer.position += 2
        kind = SyntaxKind.BangEqualsToken
      else:
        lexer.position += 1
        kind = SyntaxKind.BangToken
    of '=':
      if lexer.peek(1) == '=':
        lexer.position += 2
        kind = SyntaxKind.EqualsEqualsToken
    of '&':
      if lexer.peek(1) == '&':
        lexer.position += 2
        kind = SyntaxKind.AmpersandAmpersandToken
    of '|':
      if lexer.peek(1) == '|':
        lexer.position += 2
        kind = SyntaxKind.PipePipeToken
    else:
      discard

  if kind == SyntaxKind.BadToken:
    lexer.mDiagnostics.add(fmt"ERROR: Bad character input: '{lexer.current}'")
    lexer.position += 1

  if text.len == 0:
    text = lexer.text[start..<lexer.position]

  return newToken(kind, start, text, value)

proc diagnostics*(lexer: Lexer): seq[string] =
  lexer.mDiagnostics
